### PR TITLE
Run criteria multipiping

### PR DESCRIPTION
**Pull request by Ratysz**
# Objective

- Implement piping of results from an arbitrary amount (well, up to 16) of run criteria to enable constructing proper criteria combinators.
- Sneak in "Optional `.system()`, part 5 (criteria piping)", shhh.

## Solution

- Rework `RunCriteriaContainer` to store a `RunCriteriaTrait` trait object instead of the single/piped enum.
- Implement pseudovariadics codegen for the aforementioned trait object and piping.
- Expand `parallel_run_criteria` test to cover the new cases; expand documentation of `RunCriteria::pipe()`.

## Relevant issues
- #1295
- #2233

---

Quick example:
```rust
stage.with_system_run_criteria(
    ("every second time", "every third time")
        .pipe(|input: In<(ShouldRun, ShouldRun)>| match input {
            In((ShouldRun::Yes, ShouldRun::Yes)) => ShouldRun::Yes,
            _ => ShouldRun::No,
        })
        .label("every sixth time"),
)
```

~~I'll see if I can do something about that double indirection (the `RunCriteriaInner` thing in the implementation).~~ (Forgot how Rust works for a moment there.)

### CODE REVIEW

**Comment by mockersf** [comment]
wouldn't that be a join rather than a pipe ?

**Comment by Ratysz** [comment]
Hm, I can see it like that. Although, as per [previous discussion](https://github.com/bevyengine/bevy/pull/1675#issuecomment-803007472), I think it's pretty important to highlight the, well, piping of outputs into an input, rather than the order of operations alone.

**Comment by Ratysz** [comment]
bors try

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 1]
```diff
@@ -111,59 +106,154 @@ impl RunCriteriaContainer {
...
-            RunCriteriaInner::Single(system) => system.name(),
-            RunCriteriaInner::Piped { system, .. } => system.name(),
-        }
+    pub fn update_parent_indices(&mut self, labels: &HashMap<BoxedRunCriteriaLabel, usize>) {
+        self.parents = self
+            .after
+            .iter()
+            .take(self.inner.parents_len())
+            .map(|label| {
+                *labels.get(label).unwrap_or_else(|| {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:124]: ~~Nit: can't we just use `expect` here?~~

Nope, apparently this way dodges string formatting costs: https://stackoverflow.com/a/63291643/3234149

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Great code quality, as always. The multi-piping API is nice (and a very natural extension), the system yeet is 💯.

**Comment by alice-i-cecile** [comment]
Adding to 0.6 milestone for the optional .system aspects. If the rest of the PR is controversial, we can split this out.

**Review by wilk10 — APPROVED** [review]
This is excellent and enables some nice use cases, thanks a lot!

**Comment by cart** [comment]
Removed from the 0.6 milestone in the interest of time. Pipes are pretty niche, so requiring `.system()` in those cases doesn't bother me.

**Comment by cart** [comment]
Pulling in https://github.com/bevyengine/rfcs/pull/45 participants: @alice-i-cecile @DJMcNab @inodentry @maniwani @hymm
I haven't done a full review of the stageless RFC yet: how does "multipiping" play into those plans?

**Comment by alice-i-cecile** [comment]
> I haven't done a full review of the stageless RFC yet: how does "multipiping" play into those plans?

The current plans there are that run criteria automatically compose with AND.

Run criteria can no longer loop, so this simple rule is sufficient (at least for a first attempt).

**Comment by cart** [comment]
Testing @bevyengine/ecs-team pings

**Comment by maniwani** [comment]
> I haven't done a full review of the stageless RFC yet: how does "multipiping" play into those plans?

For some more clarity, in current Bevy, RC can be referenced by label (like any other system) and they're collectively evaluated at the beginning of each stage.

Both of those properties go away in the RFC. Instead, each system gets a unique instance of any RC it's been assigned. And the new definition is that a system's RC must return `true` "just before" it runs.  RFC also limits RC to read-only data access.

Current plan is to have the executor evaluate all of a system's RC inline (in the main executor task), just after that system becomes ready (no remaining dependencies). We'd only spawn a task for a system if all its RC returned `true`, otherwise we just mark it as completed. (When considering a system for access conflicts, we include the access of its RC.)

For guarding a group of systems, labels (which conceptually map to sub-graphs in the overall system graph) can also be assigned RC. When the executor encounters a ready system with "not-yet-encountered" labels, it would evaluate the label run criteria first. All systems with labels whose RC were evaluated and returned `false` would be immediately marked as completed.

**Comment by alice-i-cecile** [comment]
I'm going to close this out: I think it's increasingly unlikely that we go with this particular design and #4391 solves the underlying problem in a much more elegant way.

And with `iyes_loopless` out, there's much less need for an urgent solution.

**Pull request closed without merging**